File: java/jimmer-sql-graphql/repository/src/main/java/org/babyfish/jimmer/sql/example/repository/AuthorRepository.java
package org.babyfish.jimmer.sql.example.repository;

import org.babyfish.jimmer.spring.repo.support.AbstractJavaRepository;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.Predicate;
import org.babyfish.jimmer.sql.example.model.Author;
import org.babyfish.jimmer.sql.example.model.AuthorTable;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public class AuthorRepository extends AbstractJavaRepository<Author, Long> {

    private static final AuthorTable table = AuthorTable.$;

    protected AuthorRepository(JSqlClient sql) {
        super(sql);
    }

    public List<Author> findByName(String name, Fetcher<Author> fetcher) {
        return sql
                .createQuery(table)
                .whereIf(
                        name != null,
                        () -> Predicate.or(
                                table.firstName().eq(name),
                                table.lastName().eq(name)
                        )
                )
                .select(table.fetch(fetcher))
                .execute();
    }
}

File: kotlin/jimmer-sql-graphql-kt/repository/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/repository/AuthorRepository.kt
package org.babyfish.jimmer.sql.example.kt.repository

import org.babyfish.jimmer.spring.repo.support.AbstractKotlinRepository
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.spring.repository.orderBy
import org.babyfish.jimmer.sql.example.kt.model.Author
import org.babyfish.jimmer.sql.example.kt.model.firstName
import org.babyfish.jimmer.sql.example.kt.model.lastName
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.or
import org.babyfish.jimmer.sql.kt.ast.query.specification.KSpecification
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Repository

@Repository
class AuthorRepository(
    sql: KSqlClient
) : AbstractKotlinRepository<Author, Long>(sql) {

    fun findByName(name: String?, fetcher: Fetcher<Author>?): List<Author> =
        sql.executeQuery(Author::class) {
            name?.let {
                where(
                    or(
                        table.firstName eq it,
                        table.lastName eq it
                    )
                )
            }
            select(table.fetch(fetcher))
        }
}
